### ISSUE TITLE
Feature request: decoder layer hook API for inference time activation engineering

### ISSUE THREAD

**Issue by trevorgordon981** [Blaizzy/mlx-vlm#998]
## Description

mlx_vlm currently has no mechanism to intercept and modify the residual stream at decoder layer boundaries during inference. This makes it impossible to implement activation engineering techniques (abliteration, steering vectors, representation engineering) without directly editing mlx_vlm source files.

I am requesting an official hook or callback API that allows users to register functions that run after each decoder layer's forward pass, with access to the residual stream tensor. This would make inference time interventions a first class feature.

## Use Case

I run Qwen3.5 397B A17B (4bit) on a Mac Studio M3 Ultra via mlx_vlm for a personal AI assistant. The model ships with PRC aligned content policies that I want to remove using abliteration (projecting refusal directions out of the residual stream at targeted layers).

The technique requires intercepting the hidden state `h` after each decoder layer and subtracting the component that points along precomputed refusal directions. This is a simple dot product and subtraction per targeted layer, adding negligible overhead.

Currently, the only way to do this in mlx_vlm is to edit the source of `models/qwen3_5/language.py` directly, inserting projection code into the layer loop inside `Qwen3_5Model.__call__`. This patch breaks on every `pip install --upgrade mlx-vlm`.

Monkey patching `DecoderLayer.__call__` does not work because mlx's `nn.Module.__call__` is implemented in C++ and does not resolve Python level method overrides on subclasses (see ml-explore/mlx#3387).

## Proposed API

A configuration based approach that requires no code changes from users:

```json
// ~/.config/mlx-vlm/layer_hooks.json or passed via --layer-hooks
{
  "post_layer_projections": [
    {
      "direction_file": "/path/to/refusal_direction.npz",
      "top_k": 16
    },
    {
      "direction_file": "/path/to/safety_direction.npz",
      "top_k": 16
    }
  ]
}
```

The `.npz` files contain:
- `directions`: array of shape `[num_layers, hidden_dim]` (unit vectors)
- `magnitudes`: array of shape `[num_layers]` (used to select top k layers)

On model load, mlx_vlm reads the config, loads the direction arrays, selects the top k layers by magnitude, and inserts projection operations into the layer loop. The projection for each targeted layer is:

```python
for d in directions_for_this_layer:
    d_cast = d.astype(h.dtype)
    h = h - mx.sum(h * d_cast, axis=-1, keepdims=True) * d_cast
```

Alternatively, a more general callback API:

```python
# Register a hook that fires after each decoder layer
def my_hook(layer_idx: int, hidden_states: mx.array) -> mx.array:
    # modify and return hidden_states
    return hidden_states

server.register_post_layer_hook(my_hook)
```

## Why This Matters

Abliteration and activation engineering are becoming standard techniques for customizing open weight models. FailSpy's original work demonstrated that refusal is a direction in activation space that can be projected out at inference time with zero weight modification. My own research (documented at https://alooftwaffle.substack.com/p/what-happens-when-you-abliterate) extended this to MoE architectures and found that inference time hooks are strictly more effective than weight baking on MoE models.

mlx_vlm is the best option for serving large VLMs on Apple Silicon, but without hook support, anyone doing activation engineering has to maintain fragile source patches. An official API would make this sustainable.

## Current Workaround

Editing `mlx_vlm/models/qwen3_5/language.py` to add projection code directly in the `Qwen3_5Model.__call__` layer loop. This is the only approach that works because:
1. Python monkey patching is ignored by mlx's C++ `nn.Module.__call__` dispatch
2. The existing `_attach_abliteration` hook in `server.py` uses monkey patching and therefore silently fails
3. Weight baking is less effective on MoE models (see the Substack article above for empirical evidence)

## Environment

- mlx_vlm version: 0.4.4
- Model: Qwen3.5 397B A17B 4bit (MoE, 60 decoder layers, 4096 hidden dim)
- macOS, Apple Silicon M3 Ultra 512GB
